mgr: Add one finisher thread per module

kotreshhr opened the pull request:
Adds a finisher thread for each module. Each module commands are
executed via corresponding finisher thread. There is generic
finisher thread via which all other commands like notify, config
change are run.

Fixes: https://tracker.ceph.com/issues/51177
Signed-off-by: Kotresh HR <khiremat@redhat.com>





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

vshankar commented:
@kotreshhr @batrick I'm _not_ against the approach taken, however, I would like to check if the point made here(*) is worth considering?

(*): https://tracker.ceph.com/issues/51177#note-4

batrick commented on src/mgr/Mgr.cc:
```diff
@@ -57,17 +57,27 @@ Mgr::Mgr(MonClient *monc_, const MgrMap& mgrmap,
+    if (thread_name.size() > 15) {
+        thread_name.resize(15);
+    }
+    finishers.emplace(module_name, std::make_shared<Finisher>(g_ceph_context, "Mgr", thread_name));
```
Why not make the finisher  part of the module class?

kotreshhr replied:
Do you mean part of `PyModuleRegistry` class ? The `finishers` is a single map which contains pointers to generic and per module finishers. The reference to this finishers map is required across different classes like `DeamonSever`, `ActivePyModules`, `PyModuleRegistry`.  The `Mgr::init()` (active_mgr) is called via `MgrStandby::handle_mgr_map()` via active mgr's finisher thread. How does it help if finisher is part of `PyModuleRegistry` class ?

batrick replied:
No, part of `ActivePyModule` (I think that's the best place). Then you can lookup the module and queue it there.

kotreshhr replied:
@batrick Done. PTAL.

badone commented:
> I'm a tiny bit worried about bloating up ceph-mgr's memory usage with this approach. It solves the issue in hand, however, we are adding up a finite memory (queue, locks, etc..) for each mgr module. IIRC, there was some discussion (somewhere) about trimming ceph-mgr's memory footprint.
> 
> @badone ^^

I don't recall. Are you sure it was me involved (did you mean Boris maybe)?

vshankar commented:
> > I'm a tiny bit worried about bloating up ceph-mgr's memory usage with this approach. It solves the issue in hand, however, we are adding up a finite memory (queue, locks, etc..) for each mgr module. IIRC, there was some discussion (somewhere) about trimming ceph-mgr's memory footprint.
> > @badone ^^
> 
> I don't recall. Are you sure it was me involved (did you mean Boris maybe)?

Maybe, I'm not sure. It was a while back...

Anyway, we need more eyes on this change.

badone commented:
@vshankar Do you see it adding significantly to the memory usage of the manager? Seems to me the additional overhead of having a finisher for each module plus a generic one shouldn't be massively different to having just a generic one since the content of the combined finisher queues should be about the same size as the queue of the finisher in the one finisher approach right?

vshankar commented:
> @vshankar Do you see it adding significantly to the memory usage of the manager? Seems to me the additional overhead of having a finisher for each module plus a generic one shouldn't be massively different to having just a generic one since the content of the combined finisher queues should be about the same size as the queue of the finisher in the one finisher approach right?

There would be one thread per mgr plugin. That's different that having one thread which is the case now. And then, there is a (finisher) thread even for modules that are not enabled (which can be worked around). But still...

How many plugins do we have now in ceph-mgr?

badone commented:
In the mid-to-high thirties if my quick count is correct but they are not all enabled/active of course.

kotreshhr commented:
> @vshankar Do you see it adding significantly to the memory usage of the manager? Seems to me the additional overhead of having a finisher for each module plus a generic one shouldn't be massively different to having just a generic one since the content of the combined finisher queues should be about the same size as the queue of the finisher in the one finisher approach right?

Yes, there is nothing changed w.r.t queue size. With this approach, the queue size is distributed among multiple threads which was with the single thread before. The added memory usage is the new Finisher object per new thread. I am not very sure yet how much that would add up to resource consumption.

kotreshhr commented:
Ran `fs` and `rados:mgr` suite.

http://pulpito.front.sepia.ceph.com/khiremat-2022-09-02_09:35:13-fs-wip-khiremat-47893-test-mgr-finisher-thread-0-distro-default-smithi/
http://pulpito.front.sepia.ceph.com/khiremat-2022-09-06_11:22:19-rados:mgr-wip-khiremat-47893-test-mgr-finisher-thread-0-distro-default-smithi/

kotreshhr commented:
Entire rados suite scheduled:
http://pulpito.front.sepia.ceph.com/khiremat-2022-09-12_10:34:08-rados-wip-khiremat-47893-test-mgr-finisher-thread-0-distro-default-smithi/

vshankar commented:
I guess this change is fine since noone has shown objection on the approach taken here.

@badone - @kotreshhr has run additionally this through rados:mgr. Would that be enough as a source of validation?

vshankar commented:
ping?

vshankar commented:
> I guess this change is fine since noone has shown objection on the approach taken here.
> 
> @badone - @kotreshhr has run additionally this through rados:mgr. Would that be enough as a source of validation?

ping?

badone commented:
> > I guess this change is fine since noone has shown objection on the approach taken here.
> > @badone - @kotreshhr has run additionally this through rados:mgr. Would that be enough as a source of validation?
> 
> ping?

Have Patrick's concerns been addressed? It's still showing a requested change.

kotreshhr commented:
> > > I guess this change is fine since noone has shown objection on the approach taken here.
> > > @badone - @kotreshhr has run additionally this through rados:mgr. Would that be enough as a source of validation?
> > 
> > 
> > ping?
> 
> Have Patrick's concerns been addressed? It's still showing a requested change.

@badone I had missed noticing Patrick comment. I am working on it. I will refresh the PR once done. Are there any additional testing that you think which needs to be run ?

badone commented:
@kotreshhr I think you can just run the same suites again on our updated PR and that should be sufficient.

kotreshhr commented:
jenkins test windows

kotreshhr commented:
Scheduled fs and rados suite:

http://pulpito.front.sepia.ceph.com/khiremat-2022-10-18_04:28:21-fs-wip-khiremat-47893-test-mgr-finisher-thread-1-distro-default-smithi/

http://pulpito.front.sepia.ceph.com/khiremat-2022-10-18_04:14:28-rados-wip-khiremat-47893-test-mgr-finisher-thread-1-distro-default-smithi/

vshankar commented:
> Scheduled fs and rados suite:
> 
> http://pulpito.front.sepia.ceph.com/khiremat-2022-10-18_04:28:21-fs-wip-khiremat-47893-test-mgr-finisher-thread-1-distro-default-smithi/

Do we need to run the entire fs suite? Wouldn't mgr/{volumes,snap_schedule,mirroring,status,mds_autoscaler} tests suffice?

kotreshhr commented:
> > Scheduled fs and rados suite:
> > http://pulpito.front.sepia.ceph.com/khiremat-2022-10-18_04:28:21-fs-wip-khiremat-47893-test-mgr-finisher-thread-1-distro-default-smithi/
> 
> Do we need to run the entire fs suite? Wouldn't mgr/{volumes,snap_schedule,mirroring,status,mds_autoscaler} tests suffice?
Wanted to run the full fs suite as the changes were in ceph-mgr code and not the plugin. The usage of ceph commands in fs suite might give it some extra coverage ?

vshankar commented:
> > > Scheduled fs and rados suite:
> > > http://pulpito.front.sepia.ceph.com/khiremat-2022-10-18_04:28:21-fs-wip-khiremat-47893-test-mgr-finisher-thread-1-distro-default-smithi/
> > 
> > 
> > Do we need to run the entire fs suite? Wouldn't mgr/{volumes,snap_schedule,mirroring,status,mds_autoscaler} tests suffice?
> 
> Wanted to run the full fs suite as the changes were in ceph-mgr code and not the plugin. The usage of ceph commands in fs suite might give it some extra coverage ?

ACK -- do share how the tests go.

batrick commented on src/mgr/ActivePyModule.h:
```diff
@@ -45,12 +46,18 @@ class ActivePyModule : public PyModuleRunner
-    : PyModuleRunner(py_module_, clog_)
-  {}
+    : PyModuleRunner(py_module_, clog_),
+      fin_thread_name("m-fin-" + thread_name),
```
what does this look like in practice? It feels strange constructing this based on the thread name.

kotreshhr replied:
thread_name is actually the module name.

kotreshhr replied:
```
  35   Thread 0x7f78dd20b700 (LWP 367526) "m-fin-balancer"  futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x55d1407de488) at ../sysdeps/nptl/futex-internal.h:183
  37   Thread 0x7f78dc1c9700 (LWP 367528) "m-fin-crash"     futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x55d1407de748) at ../sysdeps/nptl/futex-internal.h:183
  39   Thread 0x7f78db1c7700 (LWP 367530) "m-fin-devicehea" futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x55d1407dea08) at ../sysdeps/nptl/futex-internal.h:183
  40   Thread 0x7f78da1c5700 (LWP 367532) "m-fin-iostat"    futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x55d1407decc8) at ../sysdeps/nptl/futex-internal.h:183
  41   Thread 0x7f78d91c3700 (LWP 367534) "m-fin-nfs"       futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x55d1407def88) at ../sysdeps/nptl/futex-internal.h:183
  42   Thread 0x7f78d81c1700 (LWP 367536) "m-fin-orchestra" futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x55d1407df248) at ../sysdeps/nptl/futex-internal.h:183
  44   Thread 0x7f78d71bf700 (LWP 367538) "m-fin-pg_autosc" futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x55d1407df508) at ../sysdeps/nptl/futex-internal.h:183
  46   Thread 0x7f78d61bd700 (LWP 367540) "m-fin-progress"  futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x55d1407df7c8) at ../sysdeps/nptl/futex-internal.h:183
  47   Thread 0x7f78d517b700 (LWP 367542) "m-fin-status"    futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x55d1407dfa88) at ../sysdeps/nptl/futex-internal.h:183
  49   Thread 0x7f78d4179700 (LWP 367544) "m-fin-telemetry" futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x55d1407dfd48) at ../sysdeps/nptl/futex-internal.h:183

```

kotreshhr replied:
But the mgr perf dump is not picking this up for some reason. I will dig in further.

```
    "finisher-Mgr": {                   <--------------------  Generic finisher thread
        "queue_len": 0,
        "complete_latency": {
            "avgcount": 2,
            "sum": 0.114045908,
            "avgtime": 0.057022954
        }
    },
    "finisher-Mgr-fin": {         <---------------------- Per module finisher thread 1
        "queue_len": 0,
        "complete_latency": {
            "avgcount": 0,
            "sum": 0.000000000,
            "avgtime": 0.000000000
        }
    },
    "finisher-Mgr-fin-0x5634056695f0": {  <-------------------------- Per module finisher thread 2
        "queue_len": 0,
        "complete_latency": {
            "avgcount": 0,
            "sum": 0.000000000,
            "avgtime": 0.000000000
        }
    },

```

kotreshhr replied:
Ok, fin_thread_name is the actual thread name displayed in gdb, finisher name is the second arg. Fixed it.

```
    "finisher-Mgr": {
        "queue_len": 0,
        "complete_latency": {
            "avgcount": 2,
            "sum": 0.034990320,
            "avgtime": 0.017495160
        }
    },
    "finisher-MgrStandby": {
        "queue_len": 0,
        "complete_latency": {
            "avgcount": 0,
            "sum": 0.000000000,
            "avgtime": 0.000000000
        }
    },
    "finisher-balancer": {
        "queue_len": 0,
        "complete_latency": {
            "avgcount": 0,
            "sum": 0.000000000,
            "avgtime": 0.000000000
        }
    },
    "finisher-cmd_finisher": {
        "queue_len": 0,
        "complete_latency": {
            "avgcount": 9,
            "sum": 0.000637829,
            "avgtime": 0.000070869
        }
    },
    "finisher-crash": {
        "queue_len": 0,
        "complete_latency": {
            "avgcount": 0,
            "sum": 0.000000000,
            "avgtime": 0.000000000
        }
    },
    "finisher-devicehealth": {
        "queue_len": 0,
        "complete_latency": {
            "avgcount": 0,
            "sum": 0.000000000,
            "avgtime": 0.000000000
        }
    },

```

batrick replied:
Very good.

> thread_name is actually the module name.

Ok, but why not:

`fin_thread_name("m-fin-" + py_module->module_name)`

I think that looks clearer.

kotreshhr replied:
Done

batrick reviewed (changes requested):
Few requests:

* I think we also need to update `ActivePyModules::notify_all` and others to use the new per-module finisher.
* Add basic QA tests with `qa/tasks/check_counter.py` to lookat the per-module finisher stats. Should be in `rados` and `fs` suites to start.

Some other things to do not necessarily in this PR:

* perf counters for each type of event
* Formal OpTracker for all commands/notify events. Right now we have 0 introspection for what's in the various mgr queues.
* warnings when queueing events about stuck finishers (little tricky as I don't think this is done anywhere else).

kotreshhr commented:
> * I think we also need to update `ActivePyModules::notify_all` and others to use the new per-module finisher.

@batrick This adds in concurrency between module notifications and other cmds. Previously all the module notifications would get executed in a series and no other cmds in would get executed in between until all the notifications are done. With the change, a few new cmds of some modules might go through before all the module notifications are completed. Is this not a problem ? The same is the case for config notify.

batrick commented:
> > * I think we also need to update `ActivePyModules::notify_all` and others to use the new per-module finisher.
> 
> @batrick This adds in concurrency between module notifications and other cmds. Previously all the module notifications would get executed in a series and no other cmds in would get executed in between until all the notifications are done. With the change, a few new cmds of some modules might go through before all the module notifications are completed. Is this not a problem ? The same is the case for config notify.

I think that's acceptable. We don't guarantee anywhere that all modules will process notifications in lockstep.

kotreshhr commented:
Pending items

- [ ] Add basic QA tests with qa/tasks/check_counter.py to lookat the per-module finisher stats. Should be in rados and fs suites to start.

kotreshhr commented:
jenkins test make check

kotreshhr commented:
jenkins test make check arm64

kotreshhr commented:
jenkins retest this please

kotreshhr commented:
jenkins retest this please

kotreshhr commented:
> Fixed the test and triggered teuthology job:
> 
> http://pulpito.front.sepia.ceph.com/khiremat-2022-12-02_11:12:21-fs:volumes-wip-khiremat-47893-test-mgr-finisher-thread-5-distro-default-smithi/

The latest run:
http://pulpito.front.sepia.ceph.com/khiremat-2022-12-02_15:36:44-fs:volumes-wip-khiremat-47893-test-mgr-finisher-thread-5-distro-default-smithi/

kotreshhr commented on qa/suites/fs/volumes/tasks/volumes/test/per_module_finisher_stats.yaml:
```diff
@@ -0,0 +1,18 @@
+tasks:
+  - check-counter:
+      counters:
+        mgr:
```
@vshankar @batrick Any idea how can I make this run only on active `mgr`. This config is making it run on all the mgrs running including active and standby. It's failing on standby, as the modules are not loaded in standby.

vshankar replied:
Hmmm.. check_counter does try to get from all daemons. Maybe teach the task to only fetch from active daemons or run only a single instance of ceph-mgr?

kotreshhr commented:
jenkins test api

kotreshhr commented on qa/workunits/mgr/test_per_module_finisher.sh:
```diff
@@ -0,0 +1,46 @@
+ceph iostat &
+pid=$!
+sleep 3
+kill -15 $pid
```
Done

kotreshhr commented:
jenkins retest this please

kotreshhr commented:
Teuthology Tests:

fs:mgr - http://pulpito.front.sepia.ceph.com/khiremat-2022-12-08_09:24:47-fs:mgr-wip-khiremat-47893-test-mgr-finisher-thread-6-distro-default-smithi/

kotreshhr commented:
jenkins test make check

kotreshhr commented:
jenkins retest this please

batrick commented on src/mgr/ActivePyModule.h:
```diff
@@ -45,12 +46,18 @@ class ActivePyModule : public PyModuleRunner
-    : PyModuleRunner(py_module_, clog_)
-  {}
+    : PyModuleRunner(py_module_, clog_),
+      fin_thread_name("m-fin-" + py_module->get_name()),
+      finisher(g_ceph_context, thread_name, fin_thread_name.substr(0,15))
```
```suggestion
      fin_thread_name(std::string("m-fin-" + py_module->get_name()).substr(0,15)),
      finisher(g_ceph_context, thread_name, fin_thread_name)
```

batrick commented on src/mgr/ActivePyModules.cc:
```diff
@@ -544,6 +544,8 @@ void ActivePyModules::start_one(PyModuleRef py_module)
 
       dout(4) << "Starting thread for " << name << dendl;
       active_module->thread.create(active_module->get_thread_name());
+      dout(4) << "Starting per active module finisher thread " << name << dendl;
```
```suggestion
      dout(4) << "Starting active module " << name <<" finisher thread " << active_module->fin_thread_name << dendl;
```

batrick commented on src/mgr/ActivePyModules.cc:
```diff
@@ -552,6 +554,13 @@ void ActivePyModules::shutdown()
 
+  // Stop per active module finisher thread
+  for (auto& [name, module] : modules) {
+      dout(4) << "Stopping per active module finisher thread " << name << dendl;
```
```suggestion
      dout(4) << "Stopping active module " << name << " finisher thread" << dendl;
```

batrick commented on src/mgr/ActivePyModules.h:
```diff
@@ -189,6 +189,10 @@ class ActivePyModules
                   const std::string &notify_id);
   void notify_all(const LogEntry &log_entry);
 
+  auto& get_module_finisher(const std::string &name) {
```
```suggestion
  auto& get_module_finisher(std::string_view name) {
```

kotreshhr replied:
`modules` is a map. modules<std::string, std::shared_ptr<ActivePyModule>>. So it again needs to be converted to std::string which is not implicit. So let's stick to std::string reference ?

batrick replied:
sure

batrick commented on src/mgr/PyModuleRegistry.h:
```diff
@@ -227,5 +227,15 @@ class PyModuleRegistry
     return v;
   }
 
+  bool is_module_active(const std::string &name) {
```
```suggestion
  bool is_module_active(std::string_view name) {
```

kotreshhr replied:
The same as above. It needs to be converted back to std::string.

batrick commented on src/mgr/PyModuleRegistry.h:
```diff
@@ -227,5 +227,15 @@ class PyModuleRegistry
+    return active_modules->module_exists(name);
+  }
+
+  auto& get_active_module_finisher(const std::string &name) {
```
```suggestion
  auto& get_active_module_finisher(std::string_view name) {
```

kotreshhr replied:
The same as above. It needs to be converted back to std::string.

batrick commented on qa/suites/fs/mgr/per_module_finisher_thread/tasks/per_module_finisher_stats.yaml:
```diff
@@ -0,0 +1,18 @@
+            - "finisher-progress.complete_latency.avgcount"
+            - "finisher-status.complete_latency.avgcount"
+            - "finisher-telemetry.complete_latency.avgcount"
+            - "finisher-volumes.complete_latency.avgcount"
```
Since this is in the `fs` suite, I think we should just make it specific to the volumes plugin (i.e., remove these other modules).

batrick replied:
You can also specify a minimum if you know the exact number of commands that will be queued. I'd also suggest enhancing `check_counter.py` to allow specifying a max. Then you can confirm the queue is drained in this test (max == 0).

batrick replied:
youc an put the generic test in the `rados` suite

kotreshhr replied:
@batrick Done. PTAL

batrick commented on qa/suites/rados/mgr/tasks/workunits.yaml:
```diff
@@ -13,4 +13,4 @@ tasks:
         client.0:
-          - mgr
\ No newline at end of file
+          - mgr/test_localpool.sh
```
explain this change?

kotreshhr replied:
Earlier, the `test_per_module_finisher.sh` was placed under `mgr` directory where `test_localpool.sh` is present. It's not longer relevant. I will remove this change.

batrick replied:
k

kotreshhr replied:
Sorry this is still relevant for the same reason stated above. If not, this test would run both `test_localpool.sh` and `test_per_module_finisher.sh`.

batrick reviewed (changes requested):
Also, I think this simple test can just be moved to `qa/tasks/cephfs/test_volumes.py`.

I think when I initially recommended this structure that it would be a bit more complex.

kotreshhr commented on qa/tasks/check_counter.py:
```diff
@@ -50,6 +52,8 @@ def end(self):
         if cluster_name is None:
             cluster_name = next(iter(self.ctx.managers.keys()))
 
+        active_mgr = json.loads(self.mon_manager.raw_cluster_cmd("mgr", "dump", "--format=json-pretty"))["active_name"]
```
This patch is failing here in testing. It's not finding `mon_manager`. I will fix this up.

kotreshhr replied:
Fixed

kotreshhr commented:
jenkins test make check

kotreshhr commented:
jenkins test make check

kotreshhr commented:
jenkins test make check

kotreshhr commented:
@batrick PTAL

batrick commented:
@neha-ojha please have someone from RADOS take a look at this PR. It'll need QA'd by RADOS too.

badone commented:
I will take a look.

rishabh-d-dave commented:
@kotreshhr @batrick Should this PR be put through CephFS QA now?

kotreshhr commented:
> I will take another look.

@badone Did you get time to take look at this ?

badone commented:
> > I will take another look.
> 
> @badone Did you get time to take look at this ?

I'm sorry @kotreshhr Give me a little more time please.

badone commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -7888,3 +7888,29 @@ def test_unparseable_metafile_on_legacy_to_v1_upgrade(self):
+class TestPerModuleFinsherThread(TestVolumesHelper):
+    """
+    Per module finisher thread tests related to mgr/volume cmds.
+    This is used in correspondence with check_counter wiht min val being 4
```
Nit: wiht typo

kotreshhr replied:
Done

badone commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -7888,3 +7888,29 @@ def test_unparseable_metafile_on_legacy_to_v1_upgrade(self):
+class TestPerModuleFinsherThread(TestVolumesHelper):
+    """
+    Per module finisher thread tests related to mgr/volume cmds.
+    This is used in correspondence with check_counter wiht min val being 4
```
Maybe rather than correspondence you mean in parallel, in conjunction, or something similar (this is also just a nit pick)

kotreshhr replied:
Done

badone commented on src/mgr/ActivePyModules.cc:
```diff
@@ -591,7 +601,8 @@ void ActivePyModules::notify_all(const std::string &notify_type,
     // C++ code, and avoid any potential lock cycles.
     dout(15) << "queuing notify (" << notify_type << ") to " << name << dendl;
     // workaround for https://bugs.llvm.org/show_bug.cgi?id=35984
```
Should this be directly above line 605 since it refers to an issue with lambdas

kotreshhr replied:
Done

badone commented on src/mgr/ActivePyModules.cc:
```diff
@@ -615,7 +626,8 @@ void ActivePyModules::notify_all(const LogEntry &log_entry)
     // probably ephemeral.
     dout(15) << "queuing notify (clog) to " << name << dendl;
     // workaround for https://bugs.llvm.org/show_bug.cgi?id=35984
```
Same as above

kotreshhr replied:
Done

badone commented on src/mgr/ActivePyModules.cc:
```diff
@@ -1309,7 +1321,8 @@ void ActivePyModules::config_notify()
     // C++ code, and avoid any potential lock cycles.
     dout(15) << "notify (config) " << name << dendl;
     // workaround for https://bugs.llvm.org/show_bug.cgi?id=35984
```
Ditto

kotreshhr replied:
Done

badone reviewed (changes requested):
We definitely need this and it looks good to me aside from the nits I've highlighted.

badone commented:
In your  first commit message "There is generic
finisher thread via which all other commands like notify, config
change and yet all is run." what is "yet all"?

kotreshhr commented:
> In your first commit message "There is generic finisher thread via which all other commands like notify, config change and yet all is run." what is "yet all"?

It includes metric queries like `handle_osd_perf_metric_query_updated`, `handle_mds_perf_metric_query_updated`, mgr shutdown `Mgr::shutdown()`, module start `ActivePyModules::start_one

I have modified the commit message and removed that as it was necessary.

kotreshhr commented:
jenkins test make check

kotreshhr commented:
jenkins test make check arm64

vshankar reviewed (approved):
LGTM

kotreshhr commented:
@yuriw Could you please pick this up for testing and prioritize ?

vshankar commented:
> @yuriw Could you please pick this up for testing and prioritize ?

Already in @yuriw branch.

athanatos commented:
The way this PR nests rados/mgr/per_module_finisher_thread under rados/mgr doesn't work -- several things end up being duplicated/concatenated resulting in an incorrect yaml since rados/mgr contains facets for a single suite rather than a collection of independent suites.  For example:

http://qa-proxy.ceph.com/teuthology/yuriw-2023-02-22_00:06:40-rados-wip-yuri6-testing-2023-02-21-1406-distro-default-smithi/7183156/teuthology.log

```  
roles:
  - - mgr.x
    - mon.a
    - mon.c
    - mds.a
    - mds.c
    - osd.0
    - client.0
  - - mgr.y
    - mgr.z
    - mon.b
    - mds.b
    - osd.1
    - osd.2
    - osd.3
    - client.1
  - - mon.a
    - mgr.x
    - osd.0
    - osd.1
    - osd.2
    - mds.a
    - mds.b
    - client.0
```

is a concatenation of two different cluster configs.  From the description:

```
2023-02-22T22:51:11.925 DEBUG:teuthology.run:Teuthology command: teuthology --description rados/mgr/{clusters/{2-node-mgr} debug/mgr mgr_ttl_cache/disable mon_election/classic per_module_finisher_thread/{begin/{0-install 1-ceph 2-logrotate} clusters/small-cluster conf/{client mds mon osd} distro/{centos_8} mount/kclient/{mount-syntax/{v2} mount overrides/{distro/testing/k-testing ms-die-on-skipped}} objectstore/bluestore-bitmap overrides/{ignorelist_health ignorelist_wrongly_marked_down no_client_pidfile} tasks/per_module_finisher_stats} random-objectstore$/{bluestore-bitmap} supported-random-distro$/{rhel_8} tasks/failover} --archive /home/teuthworker/archive/yuriw-2023-02-22_00:06:40-rados-wip-yuri6-testing-2023-02-21-1406-distro-default-smithi/7183156 --verbose --name yuriw-2023-02-22_00:06:40-rados-wip-yuri6-testing-2023-02-21-1406-distro-default-smithi --owner scheduled_yuriw@teuthology -- /home/teuthworker/archive/yuriw-2023-02-22_00:06:40-rados-wip-yuri6-testing-2023-02-21-1406-distro-default-smithi/7183156/orig.config.yaml
```

```
...rados/mgr/{clusters/{2-node-mgr}...clusters/small-cluster
```

Can we simply put the new task in rados/mgr/tasks instead of defining an entire new suite?

athanatos reviewed (changes requested):
See comment above, added qa yamls don't quite work.

yuriw commented:
@kotreshhr pls add `needs-qa` when fixed

kotreshhr commented:
jenkins test make check

ljflores commented:
> @kotreshhr pls add `needs-qa` when fixed

+ 1

kotreshhr commented:
jenkins test api

kotreshhr commented:
> Can we simply put the new task in rados/mgr/tasks instead of defining an entire new suite?

Yes, Done. Thank you the review.  It was added as a separate suite initially as it required a single mgr cluster. I fixed it down the line but the new suite remained.

kotreshhr commented:
> @kotreshhr pls add `needs-qa` when fixed

Sure, the changes are in place but I couldn't do sanity test since teuthology is down. I will do it once it's back and add `needs-qa`. Thank you.

kotreshhr commented:
@yuriw marked `needs-qa`. The PR is ready for testing.

kotreshhr commented:
jenkins test windows

kotreshhr commented:
jenkins test windows

kotreshhr commented:
jenkins test api

kotreshhr commented:
jenkins test windows

kotreshhr commented:
jenkins test windows

kotreshhr commented:
jenkins test windows

kotreshhr commented:
@athanatos PTAL. Addressed your comments

rzarzynski commented:
Repinging @athanatos.

athanatos reviewed (approved):
The suite changes look reasonable to me now -- as long as tests pass LGTM

lxbsz commented on qa/tasks/check_counter.py:
```diff
@@ -96,7 +112,9 @@ def end(self):
-                        if val >= minval:
+                        if isinstance(minval, int) and val >= minval:
+                            seen.add(name)
+                        elif isinstance(expected_val, int) and val == expected_val:
```
@kotreshhr This caused the following errors:

```
2023-03-28T10:34:36.302 ERROR:teuthology.run_tasks:Manager failed: check-counter
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_teuthology_main/teuthology/run_tasks.py", line 192, in run_tasks
    suppress = manager.__exit__(*exc_info)
  File "/home/teuthworker/src/git.ceph.com_teuthology_main/teuthology/task/__init__.py", line 132, in __exit__
    self.end()
  File "/home/teuthworker/src/github.com_ceph_ceph-c_9d12d2cb27d4a45c58218390bfbb56820b010b77/qa/tasks/check_counter.py", line 117, in end
    elif isinstance(expected_val, int) and val == expected_val:
UnboundLocalError: local variable 'expected_val' referenced before assignment
```

More detail please see https://pulpito.ceph.com/yuriw-2023-03-28_00:15:34-fs-wip-yuri5-testing-2023-03-27-1315-distro-default-smithi/.

Could you have a look ? Thanks.

kotreshhr replied:
Thanks Xiubo. I will take a look.

kotreshhr replied:
Fixed it. Ran the failed one with the fix. It works. Hopefully this is the last failure.
http://pulpito.front.sepia.ceph.com/khiremat-2023-03-29_06:17:28-fs-wip-yuri5-testing-2023-03-27-1315-distro-default-smithi/

kotreshhr replied:
@yuriw This needs a qa re-run after this fix.

kotreshhr commented:
jenkins test make check arm64

kotreshhr commented:
jenkins test make check arm64

kotreshhr commented:
jenkins test make check arm64

kotreshhr commented:
jenkins test make check arm64

lxbsz reviewed (approved):
LGTM.

vshankar reviewed (approved):
LGTM. Nice work.
